File: codes/Garnetwzy/42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i = 0;
        int j = height.size()-1;
        int sum = 0;
        int maxLeft = 0;
        int maxRight = 0;
        while(i <= j) {
            if(height[i] < height[j]) {
                if(height[i] > maxLeft) {
                    maxLeft = height[i];
                    i++;
                }else {
                    sum += maxLeft - height[i];
                    i++;
                }
            }else{
                if(height[j] > maxRight) {
                    maxRight = height[j];
                    j--;
                }else{
                    sum += maxRight - height[j];
                    j--;
                }
            }
        }
        return sum;
    }
};

File: codes/HellDream/42.java
class Solution {
    public int trap(int[] height) {
        int max = 0;
        int leftMax =0, rightMax = 0;
        int left = 0, right = height.length-1;
        while(left<right){
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if(leftMax<rightMax){
                max += leftMax-height[left];
                left++;
            }else{
                max+= rightMax-height[right];
                right--;
            }
        }
        return max;
    }
}